corehq/apps/api/user_updates.py

```diff
@@ -194,9 +194,9 @@ def _update_location(user, location_object, user_change_logger):
             raise UpdateUserException(_('Primary location must be included in the list of locations.'))
 
         # Verify all location ids has a SQLLocation.
-        real_ids = (SQLLocation.active_objects
-                    .filter(location_id__in=location_ids, domain=user.domain)
-                    .location_ids())
+        locations = SQLLocation.active_objects.filter(location_id__in=location_ids, domain=user.domain)
+        real_ids = locations.location_ids()
```

Comment:
I think this is effectively the same as before since Django [querysets are lazy](https://docs.djangoproject.com/en/4.2/topics/db/queries/#querysets-are-lazy), and this doesn't take advantage of [QuerySet caching](https://docs.djangoproject.com/en/4.2/topics/db/queries/#caching-and-querysets).

You want to force the evaluation of the base query, in this case `locations = SQLLocation.active_objects.filter(location_id__in=location_ids, domain=user.domain)`
```suggestion
        locations = SQLLocation.active_objects.filter(location_id__in=location_ids, domain=user.domain)
        _ = list(locations)  # forces evaluation and populates locations._result_cache which can be used by the next line
        real_ids = locations.location_ids()
```

There might be a cleaner way to force evaluation of the queryset, but you want to keep `locations` as a QuerySet so that it can be used in both the context of fetching IDs specifically, and then later on to get the objects (which you would have already done when forcing the evaluation up front).

Reply:
Thank you. Updated in 23bbc659ec2c18189208150dc3599361dd95abef